virtcontainers/clh.go

```diff
@@ -0,0 +1,1316 @@
+	return clh.terminate()
+}
+
+func (clh *cloudHypervisor) fromGrpc(ctx context.Context, hypervisorConfig *HypervisorConfig, store *store.VCStore, j []byte) error {
+	return errors.New("cloudHypervisor is not supported by VM cache")
+}
+
+func (clh *cloudHypervisor) toGrpc() ([]byte, error) {
+	return nil, errors.New("cloudHypervisor is not supported by VM cache")
+}
+
+func (clh *cloudHypervisor) save() (s persistapi.HypervisorState) {
+	s.Pid = clh.state.PID
+	s.Type = string(ClhHypervisor)
+	return
+}
+
+func (clh *cloudHypervisor) load(s persistapi.HypervisorState) {
+	clh.state.PID = s.Pid
+	clh.state.VirtiofsdPID = s.VirtiofsdPid
+}
+
+func (clh *cloudHypervisor) check() error {
+	return nil
+}
+
+func (clh *cloudHypervisor) getPids() []int {
+
+	var pids []int
+	pids = append(pids, clh.state.PID)
+
+	return pids
+}
+
+//###########################################################################
+//
+// Local helper methods related to the hypervisor interface implementation
+//
+//###########################################################################
+
+func (clh *cloudHypervisor) addDevice(devInfo interface{}, devType deviceType) error {
+	span, _ := clh.trace("addDevice")
+	defer span.Finish()
+
+	var err error
+
+	switch v := devInfo.(type) {
+	case Endpoint:
+		clh.Logger().WithField("function", "addDevice").Info(fmt.Sprintf("Adding Endpoint of type %v", v))
+		clh.cliBuilder.AddNet(CLINet{
+			device: v.Name(),
+			mac:    v.HardwareAddr(),
+		})
+
+	default:
+		clh.Logger().WithField("function", "addDevice").Warn(fmt.Sprintf("Add device of type %v is not supported.", v))
+	}
+
+	return err
+}
+
+func (clh *cloudHypervisor) clhAddVsock(vs types.VSock) error {
+	clh.Logger().WithField("function", "clhAddVsock").Info("Add vsock device contextID=", vs.ContextID)
+	return nil
+}
+
+func (clh *cloudHypervisor) Logger() *log.Entry {
+	return virtLog.WithField("subsystem", "cloudHypervisor")
+}
+
+func (clh *cloudHypervisor) capabilities() types.Capabilities {
+	span, _ := clh.trace("capabilities")
+	defer span.Finish()
+
+	clh.Logger().WithField("function", "capabilities").Info("get Capabilities")
+	var caps types.Capabilities
+	return caps
+
+}
+
+func (clh *cloudHypervisor) trace(name string) (opentracing.Span, context.Context) {
+
+	if clh.ctx == nil {
+		clh.Logger().WithField("type", "bug").Error("trace called before context set")
+		clh.ctx = context.Background()
+	}
+
+	span, ctx := opentracing.StartSpanFromContext(clh.ctx, name)
+
+	span.SetTag("subsystem", "cloudHypervisor")
+	span.SetTag("type", "clh")
+
+	return span, ctx
+}
+
+func (clh *cloudHypervisor) terminate() (err error) {
+	span, _ := clh.trace("terminate")
+	defer span.Finish()
+
+	defer func() {
+		if err != nil {
+			clh.Logger().Info("Terminate Cloud Hypervisor failed")
+		} else {
+			clh.Logger().Info("Cloud Hypervisor stopped")
+			clh.state.PID = 0
+			clh.state.VirtiofsdPID = 0
+			clh.state.state = clhNotReady
```

Comment:
Thought: This might benefit from a `cloudHypervisor.reset()` helper method (which possibly calls `CloudHypervisorState.reset()` or  `NewCloudHypervisorState()` or similar). This might help avoid bugs if we ever extend `CloudHypervisorState`.